build/Dockerfile

```diff
@@ -138,32 +178,40 @@ apt-get install -y --no-install-recommends \
     protobuf-compiler
 EOF
 
-# Build dependencies with cargo chef
-COPY --from=rust-planner ${RUST_BUILD_PATH}/recipe.json .
+# Build external dependencies with cargo chef.
+COPY --from=rust-prepare ${RUST_BUILD_PATH}/recipe.json .
 RUN cargo chef cook --release --recipe-path recipe.json
 
-# Build application
+# Build application.
 COPY ./offchain/ .
-# needed to log commit hash in runtime
-COPY ./.git ../
 RUN cargo build --release
 
+# STAGE: go-builder
 #
-# Go build
-#
+# This stage builds the Go binaries.
+# First it downloads the external dependencies and then it builds the binaries.
 FROM golang:${GO_VERSION}-bookworm as go-builder
 ARG GO_BUILD_PATH
 WORKDIR ${GO_BUILD_PATH}
+
+# Download external dependencies.
+COPY go.mod .
+COPY go.sum .
+RUN go mod download
+
+# Build application.
 COPY . .
-RUN go install -ldflags "-s -w" ./cmd/cartesi-rollups-node
+RUN go build -ldflags "-s -w" ./cmd/cartesi-rollups-node
 
+# STAGE: rollups-node
 #
-# Runtime
-#
+# This stage copies the binaries from the previous stages.
 FROM cartesi/server-manager:${SERVER_MANAGER_VERSION} as rollups-node
```

Comment:
I'm uncomfortable with this; maybe we could make some changes now.

We should have more control over the runtime image, having all we need explicit in this file.

When we use `FROM server-manager` we get another `FROM machine-emulator` and it's difficult to track what's inside those images.

My suggestion is to use `FROM debian:bookworm-slim` (or even use the sha256 hash for reproducibility and security) and copy server-manager binaries and install the machine-emulator .deb package.

Something like this should be enough, and check for library dependencies.

NOT TESTED CODE
```
FROM debian:bookworm-slim
...
COPY --from=cartesi/server-manager:${SERVER_MANAGER_VERSION} /usr/bin/server-manager /usr/bin/server-manager
...
RUN <<EOF
wget https://github.com/cartesi/machine-emulator/releases/download/$MACHINE_EMULATOR_VERSION/cartesi-machine-v$MACHINE_EMULATOR_VERSION_$ARCH.deb
apt install -y cartesi-machine-v$MACHINE_EMULATOR_VERSION_$ARCH.deb
rm cartesi-machine-v$MACHINE_EMULATOR_VERSION_$ARCH.deb
EOF
....
RUN <<EOF
addgroup --system --gid 102 adduser --system --uid 102 --ingroup cartesi --disabled-login --no-create-home --home /nonexistent --gecos "cartesi user" --shell /bin/false cartesi
```

Reply:
I'll send a PR over to this branch for your consideration.

Reply:
https://github.com/cartesi/rollups-node/pull/263

Reply:
Thanks. I'll use this PR as inspiration to create a separate stage to be the base image. This image should be used when building the Cartesi machine for the `rollups-node-snapshot` target.